docs/decisions/0015-expiration-improvements.rst

```diff
@@ -0,0 +1,94 @@
+0015 Expiration and Cancellation Data Improvements
+**************************************************
+
+Status
+======
+Proposed - December 2023
+
+Context
+=======
+There are subtleties around the Assignment cancellation and expiration user
+experience in the enterprise learner-portal frontend that require a more nuanced
+and descriptive approach to the data we persist around these states, and
+how that data is represented in API response payloads for consumption
+by the learner-portal frontend.
+
+How expiration currently works
+------------------------------
+There's currently a management command (run via a cron) that toggles the state
+of Assignments to ``CANCELLED`` under the following conditions:
+
+1. The current date is more than 90 days after the *creation* date of the assignment. We use the creation
+   date as a proxy for the date at which an assignment was allocated, which is problematic under
+   the edge case where an assignment is allocated, cancelled, and re-allocated later.
+2. The current date is greater than the inferred enrollment deadline for the assigned course.
+3. The current date is greater than the expiration date of the subsidy associated with the Assignment's
+   access policy record.
+
+Learners are notified of expiration and cancellation via the learner-portal frontend
+------------------------------------------------------------------------------------
+The learner-portal notifies learners of assignment cancellation and expiration.  These notifications
+are required to be dismissable, which means we need to persist state about dismissal
+for assignments in the backend.
+
+Decision
+========
+
+``EXPIRED`` should be a distinct Assignment State
+-------------------------------------------------
+This will help disambiguate behavior and simplify/remove business logic
+from the learner-portal frontend.  We may wish to also store the expiration reason
+for an assignment (i.e. which of the 3 cases above caused the assignment to become expired).
+
+Persist the allocation, cancellation, and expiration dates
```

Comment:
[clarification] At what point would these fields be persisted/updated? For example, if the subsidy expiration date changes, what is the trigger to update the associated assignment records' persisted expiration date, if needed?

Reply:
I should clarify those field names and intented semantics.  The fields should really correspond to the datetime at which the assignment record's `state` changed.  For example, when the record is born/allocated, we'd have `state='allocated'` and `allocated_at='2023-12-19T14:28:00Z'`.  When the assignment moves to `expired`, we'd have `state='expired'` and `expired_at='2023-12-25T00:00:00'`.